bee-network/bee-autopeering/src/task.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    peer::{
+        lists::{ActivePeersList, ReplacementList},
+        PeerStore,
+    },
+    time::SECOND,
+};
+
+use priority_queue::PriorityQueue;
+use tokio::{sync::oneshot, task::JoinHandle, time};
+
+use std::{collections::HashMap, future::Future, time::Duration};
+
+pub(crate) const MAX_SHUTDOWN_PRIORITY: u8 = 255;
+const SHUTDOWN_TIMEOUT_SECS: Duration = Duration::from_secs(5 * SECOND);
+
+pub(crate) type ShutdownRx = oneshot::Receiver<()>;
+type ShutdownTx = oneshot::Sender<()>;
+
+pub(crate) type Repeat<T> = Box<dyn for<'a> Fn(&'a T) + Send>;
+
+/// Represents types driving an event loop.
+#[async_trait::async_trait]
+pub(crate) trait Runnable {
+    const NAME: &'static str;
+    const SHUTDOWN_PRIORITY: u8;
+
+    type ShutdownSignal: Future + Send + Unpin + 'static;
+
+    async fn run(self, shutdown_rx: Self::ShutdownSignal);
+}
+
+pub(crate) struct TaskManager<S: PeerStore, const N: usize> {
+    shutdown_handles: HashMap<String, JoinHandle<()>>,
+    shutdown_order: PriorityQueue<String, u8>,
+    shutdown_senders: HashMap<String, ShutdownTx>,
```

Comment:
I'd merge handles and senders with a tuple, more light on the RAM

Reply:
We are talking about a few bytes (<100) saved here compared to Gigabytes (>= 1.000.000.000) of available RAM on servers where the nodes are running on. Number of tasks won't increase anymore either. It's not worth it IMO, and I am not a big fan of tuples either as they make everything shitty to read.

Reply:
It's also less hashing because you insert/get together. I think they make sense together since they are always bound one to each other. You could do a struct if you don't like tuples. But yeah, your call.

Reply:
They are not bound together during removal.